File: test/test_backend.py
#!/usr/bin/env python

## $Id$

# End to end test.  Tests make_work, feeder, scheduling server, client,
# file_upload_handler, validator, assimilator, timeout_check, and file_deleter
# on a large batch of workunits.  Confirms that credit is correctly granted
# and that unneeded files are deleted

from boinc import *
from test_uc import *
import time, os

class ProjectBackend(ProjectUC):
    def __init__(self, num):
        self.num = num
        ProjectUC.__init__(self, redundancy = 5, short_name = 'test_backend')
    def run(self):
        self.install()
        self.sched_install('make_work', cushion=self.num-1)
        self.start_servers()

        # wait for 500 results to be generated
        n = 0
        db = self.db_open()
        while n < self.num:
            n = num_wus_left(db)
            verbose_echo(1, "Generating results [%d/%d]" % (n,self.num));
            time.sleep(.1)
        db.close()
        verbose_echo(1, "Generating results [%d/%d]: done." % (self.num,self.num));

        # Stop the project, deinstall make_work, and install the normal
        # backend components
        self.stop()
        self.sched_uninstall('make_work')
        self.sched_install('assimilator')
        self.sched_install('file_deleter')
        self.sched_install('validate_test')
        self.sched_install('feeder')
        # self.sched_install('timeout_check', nredundancy=0)
        self.start_servers()

    def check(self):
        # Give the server 30 seconds to finish assimilating/deleting
        # TODO: use wait on all processes.
        verbose_sleep("Sleeping to allow server daemons to finish", 30)
        self.check_results(ResultUC(), self.num)

if __name__ == '__main__':
    num = sys.argv[1:] and get_int(sys.argv[1]) or 100
    test_msg("entire backend (with %d results)" % num);
    ProjectBackend(num)
    run_check_all()

File: test/test_backend.php
#! /usr/local/bin/php
<?php
    // End to end test.  Tests make_work, feeder, scheduling server, client,
    // file_upload_handler, validator, assimilator, and file_deleter on a
    // large batch of workunits.  Confirms that credit is correctly granted
    // and that unneeded files are deleted

    include_once("test.inc");

    $project = new Project;
    $user = new User();
    $host = new Host($user);
    $app = new App("upper_case");
    $app_version = new App_Version($app);

    $work = new Work($app);
    $work->wu_template = "uc_wu";
    $work->result_template = "uc_result";
    $work->redundancy = 4;
    $work->delay_bound = 86400;
    array_push($work->input_files, "input");

    $project->add_user($user);
    $project->add_app($app);
    $project->add_app_version($app_version);
    $project->install();      // must install projects before adding to hosts
    $project->install_make_work($work, 100, 5);

    $host->log_flags = "log_flags.xml";
    $host->add_user($user,$project);
    $host->install();

    echo "adding work\n";

    $work->install($project);

    $project->start_servers();

    // Start by generating a batch of 100 results
    echo "Generating 100 results... ";
    while( $project->num_wus_left() < 100 ) sleep(1);
    echo "done.\n";

    // Stop the project, deinstall make_work, and install the normal backend components
    $project->stop();
    $project->deinstall_make_work();
    $project->install_assimilator($app);
    $project->install_file_delete();
    $project->install_validate($app, 5);
    $project->install_feeder();

    while (($pid=exec("pgrep -n make_work")) != null) sleep(1);

    // Restart the server
    $project->restart();
    $project->start_servers();

    // Run the client until there's no more work
    $host->run("-exit_when_idle -skip_cpu_benchmarks");

    sleep(5);

    // *** DO TESTS HERE
    $result->state = RESULT_STATE_DONE;
    $result->exit_status = 0;
    $project->check_results(101, $result);

    // Stop the server
    $project->stop();
?>
